pkg/loop/internal/plugin_service.go

```diff
@@ -212,11 +209,43 @@ func (s *pluginService[P, S]) closeClient() (err error) {
 	return
 }
 
-func (s *pluginService[P, S]) wait(ctx context.Context) error {
+func (s *PluginService[P, S]) Wait(ctx context.Context) error {
 	select {
 	case <-ctx.Done():
 		return context.Cause(ctx)
 	case <-s.serviceCh:
 		return nil
+	case <-s.stopCh:
+		return nil
+	}
+}
+
+// XXXTestHook returns a TestPluginService.
+// It must only be called once, and before Start.
+func (s *PluginService[P, S]) XXXTestHook() TestPluginService[P, S] {
```

Comment:
I've spent a bit of time thinking about this, what do you think of adding this hook to a file with a test build tag, and then running our tests with -tags=test? (A default test build tag has been suggested before, but not accepted: https://github.com/golang/go/issues/21360)

Note: this issue is documented here https://github.com/golang/go/issues/39565, and the recommendation from go developers was to move this suggestion into a common testing package and export the shared functionality which is basically what I did.

Reply:
I don't think it is worth introducing tags for. That was a major hassle in the core repo that we only just moved away from.

> and the recommendation from go developers was to move this suggestion into a common testing package

That is what I was going to suggest. We currently only use an internal package here, with both product and test code. Using internal/test as well (or some new test package) would separate and encapsulate the concerns. However, it might require another shared package below this level, or just moving this type to that other package. For example
- move this type to `internal/plugin.Service`
- add `internal/plugintest.Service` which embeds `Service` and adds these methods
- tests can choose to import `plugintest` in place of plugin to get the extended `Service`

Reply:
Oh I think that the way that we embed this type makes it more complicated :thinking: